src/app/shared/services/base.service.spec.ts

```diff
@@ -0,0 +1,94 @@
+import { TestBed } from '@angular/core/testing';
+import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
+
+import { environment } from '../../../environments/environment';
+import { BaseService } from './base.service';
+import { HttpErrorResponse } from '@angular/common/http';
+
+describe('BaseService', () => {
+  let service: BaseService;
+  let httpMock: HttpTestingController;
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      imports: [
+        HttpClientTestingModule
+      ]
+    });
+    service = TestBed.inject(BaseService);
+    httpMock = TestBed.inject(HttpTestingController);
+  });
+
+  it('creates the service', () => {
+    expect(service).toBeTruthy();
+  });
+
+  describe('#post', () => {
+    describe('Given a valid data', () => {
+      it('returns an Observable<any>', () => {
+        const mockResponse = {
+          data: {
+            id: 10,
+            type: 'token',
+            attributes: {
+              access_token: 'lbxD2K2BjbYtNzz8xjvh2FvSKx838KBCf79q773kq2c',
+              token_type: 'Bearer',
+              expires_in: 7200,
+              refresh_token: '3zJz2oW0njxlj_I3ghyUBF7ZfdQKYXd2n0ODlMkAjHc',
+              created_at: 1597169495
+            }
+          }
+        };
+
+        const deserializedResponse = {
+          id: 10,
+          accessToken: 'lbxD2K2BjbYtNzz8xjvh2FvSKx838KBCf79q773kq2c',
+          tokenType: 'Bearer',
+          expiresIn: 7200,
+          refreshToken: '3zJz2oW0njxlj_I3ghyUBF7ZfdQKYXd2n0ODlMkAjHc',
+          createdAt: 1597169495
+        };
+
+        const data = {
+          grant_type: 'password',
+          email: 'dev@nimblehq.co',
+          password: '12345678',
+          client_id: environment.apiClientID,
+          client_secret: environment.apiClientSecret
+        };
+
+        service.post('oauth/token', data).subscribe(response => {
+          expect(response).toEqual(deserializedResponse);
+        });
+
+        const request = httpMock.expectOne(`${environment.apiBaseUrl}/api/${environment.apiVersion}/oauth/token`);
+        expect(request.request.method).toBe('POST');
+        request.flush(mockResponse);
+      });
+    });
+
+    describe('Given an INVALID data', () => {
+      it('throws error', () => {
+        const data = {
+          grant_type: 'password',
+          email: 'invalid_email@nimblehq.co',
+          password: '12345678',
+          client_id: environment.apiClientID,
+          client_secret: environment.apiClientSecret
+        };
+
+        service.post('oauth/token', data).subscribe(
+          _ => fail('Should have failed with 400 error'),
```

Comment:
Will that be the error message if this expectation is not met?

Reply:
@olivierobert  Yes the error message on testing would look like this
![Screen Shot 2563-12-10 at 12 06 57](https://user-images.githubusercontent.com/1772999/101724118-5e0e1a80-3ae0-11eb-9413-3446126053ac.png)